plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt

```diff
@@ -116,6 +148,153 @@ class SerializableIrGenerator(
+                irGet(seenVars[0]),
+                irInt(goldenMask),
+                getSerialDescriptorExpr(),
+                typeHint = compilerContext.irBuiltIns.unitType
+            )
+
+            fieldsMissedTest = irNotEquals(
+                irInt(goldenMask),
+                irBinOp(
+                    OperatorNameConventions.AND,
+                    irInt(goldenMask),
+                    irGet(seenVars[0])
+                )
+            )
+        } else {
+            val goldenMaskList = getGoldenMaskList()
+
+            var compositeExpression: IrExpression? = null
+            for (i in goldenMaskList.indices) {
+                val singleCheckExpr = irNotEquals(
+                    irInt(goldenMaskList[i]),
+                    irBinOp(
+                        OperatorNameConventions.AND,
+                        irInt(goldenMaskList[i]),
+                        irGet(seenVars[i])
+                    )
+                )
+
+                compositeExpression = if (compositeExpression == null) {
+                    singleCheckExpr
+                } else {
+                    irBinOp(
+                        OperatorNameConventions.OR,
+                        compositeExpression,
+                        singleCheckExpr
+                    )
+                }
+            }
+
+            fieldsMissedTest = compositeExpression!!
+
+            throwErrorExpr = irBlock {
+                +irInvoke(
+                    null,
+                    throwMissedFieldExceptionArrayFunc,
+                    createPrimitiveArrayOfExpression(compilerContext.irBuiltIns.intType, goldenMaskList.indices.map { irGet(seenVars[it]) }),
+                    createPrimitiveArrayOfExpression(compilerContext.irBuiltIns.intType, goldenMaskList.map { irInt(it) }),
+                    getSerialDescriptorExpr(),
+                    typeHint = compilerContext.irBuiltIns.unitType
+                )
+            }
+        }
+
+        +irIfThen(compilerContext.irBuiltIns.unitType, fieldsMissedTest, throwErrorExpr)
+    }
+
+    private fun IrBlockBodyBuilder.getSerialDescriptorExpr(): IrExpression {
+        return if (serializableDescriptor.shouldHaveGeneratedSerializer && staticDescriptor) {
+            val serializer = serializableDescriptor.classSerializer!!
+            val serialDescriptorGetter = compilerContext.referenceClass(serializer.fqNameSafe)!!.getPropertyGetter(SERIAL_DESC_FIELD)!!
+            irGet(
+                serialDescriptorGetter.owner.returnType,
+                irGetObject(serializer),
+                serialDescriptorGetter.owner.symbol
+            )
+        } else {
+            irGetField(null, generateStaticDescriptorField())
+        }
+    }
+
+    private fun IrBlockBodyBuilder.generateStaticDescriptorField(): IrField {
+        val serialDescItType = serialDescClass.defaultType.toIrType()
+
+        val function = irClass.createInlinedFunction(
```

Comment:
Why there is a need in inlined function?

Reply:
1. I think it's good not to clutter up metadata of result class
2. I not found other good way to create clinit block on java platform - it's a good way to init field in <clinit>